hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/operations/FinalizeJoinOp.java

```diff
@@ -144,22 +155,35 @@ protected void writeInternalImpl(ObjectDataOutput out) throws IOException {
         out.writeLong(clusterStartTime);
         out.writeUTF(clusterState.toString());
         out.writeObject(clusterVersion);
+        if (clusterVersion.isGreaterOrEqual(V3_9)) {
+        boolean hasPreJoinOp = preJoinOp != null;
```

Comment:
I think `hasPreJoinOp` not needed when you write via `out.writeObject()`, it should handle `null` arguments.